eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxosSpec.hs

```diff
@@ -612,7 +536,222 @@ govPolicySpec = do
+costModelsSpec =
+  describe "PlutusV3 Initialization" $ do
+    it "Updating CostModels with alwaysFails govPolicy does not validate" $ do
+      -- no initial PlutusV3 CostModels
+      modifyPParams $ ppCostModelsL .~ testingCostModels [PlutusV1 .. PlutusV2]
+
+      (dRep, committeeMember, _) <- electBasicCommittee
+      anchor <- arbitrary
+      govIdConstitution1 <-
+        enactConstitution SNothing (Constitution anchor SNothing) dRep committeeMember
+      -- propose and enact PlutusV3 Costmodels
+      govIdPPUpdate1 <-
+        enactCostModels SNothing (testingCostModels [PlutusV3]) dRep committeeMember
+
+      let alwaysFailsSh = hashPlutusScript (alwaysFails2 SPlutusV3)
+      void $
+        enactConstitution
+          (SJust (GovPurposeId govIdConstitution1))
+          (Constitution anchor (SJust alwaysFailsSh))
+          dRep
+          committeeMember
+
+      impAnn "Fail to update V3 Costmodels" $ do
+        let pparamsUpdate = def & ppuCostModelsL .~ SJust (testingCostModels [PlutusV3])
+        let govAction = ParameterChange (SJust govIdPPUpdate1) pparamsUpdate (SJust alwaysFailsSh)
+        rewardAccount <- registerRewardAccount
+        pp <- getsNES $ nesEsL . curPParamsEpochStateL
+        let proposal =
+              ProposalProcedure
+                { pProcReturnAddr = rewardAccount
+                , pProcGovAction = govAction
+                , pProcDeposit = pp ^. ppGovActionDepositL
+                , pProcAnchor = anchor
+                }
+        let tx = mkBasicTx mkBasicTxBody & bodyTxL . proposalProceduresTxBodyL .~ [proposal]
+        expectPhase2Invalid tx
+
+    it "Updating CostModels with alwaysSucceeds govPolicy but no PlutusV3 CostModels fails" $ do
+      modifyPParams $ ppCostModelsL .~ testingCostModels [PlutusV1 .. PlutusV2]
+
+      (dRep, committeeMember, _) <- electBasicCommittee
+      anchor <- arbitrary
+      let alwaysSucceedsSh = hashPlutusScript (alwaysSucceeds2 SPlutusV3)
+      void $
+        enactConstitution
+          SNothing
+          (Constitution anchor (SJust alwaysSucceedsSh))
+          dRep
+          committeeMember
+
+      let pparamsUpdate = def & ppuCostModelsL .~ SJust (testingCostModels [PlutusV3])
+      let govAction = ParameterChange SNothing pparamsUpdate (SJust alwaysSucceedsSh)
+
+      submitFailingGovAction govAction [injectFailure $ CollectErrors [NoCostModel PlutusV3]]
+
+    it "Updating CostModels and setting the govPolicy afterwards succeeds" $ do
+      modifyPParams $ ppCostModelsL .~ testingCostModels [PlutusV1 .. PlutusV2]
+
+      (dRep, committeeMember, _) <- electBasicCommittee
+      anchor <- arbitrary
+      govIdConstitution1 <-
+        enactConstitution SNothing (Constitution anchor SNothing) dRep committeeMember
+
+      let guessTheNumberSh = hashPlutusScript (guessTheNumber3 SPlutusV3)
+
+      impAnn "Minting token fails" $ do
+        tx <- mintingTokenTx @era (mkBasicTx @era mkBasicTxBody) guessTheNumberSh
```

Comment:
These type annotations are not necessary. But it is ok to keep them around, so no need to change it, just pointing it out.
```suggestion
        tx <- mintingTokenTx (mkBasicTx mkBasicTxBody) guessTheNumberSh
```